pkg/app/piped/imageprovider/gcr/gcr.go

```diff
@@ -82,25 +91,45 @@ func (p *Provider) ParseImage(image string) (*model.ImageName, error) {
 	}, nil
 }
 
-func (p *Provider) GetLatestImage(ctx context.Context, image *model.ImageName) (*model.ImageRef, error) {
-	repository, err := client.NewRepository(image, p.baseURL.String(), p.transport)
+func (g *GCR) GetLatestImage(ctx context.Context, image *model.ImageName) (*model.ImageRef, error) {
+	repo, err := name.NewRepository(image.String())
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%s is invalid repository: %w", image, err)
+	}
+	options := []google.ListerOption{
+		google.WithContext(ctx),
 	}
-	// TODO: Stop listing all tags
-	_, err = repository.Tags(ctx).All(ctx)
+	if g.authenticator != nil {
+		options = append(options, google.WithAuth(g.authenticator))
+	}
+	// TODO: Use pagination to retrieve image tags from GCR
+	// Currently, the result could be quite large size if there are a lot of tags.
+	// "google/go-containerregistry" doesn't provide any option to paginate.
+	// We can propose it to them, or just borrow and modify for us.
+	// See more: https://docs.docker.com/registry/spec/api/#listing-image-tags
+	res, err := google.List(repo, options...)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to list tags: %w", err)
+	}
+	if len(res.Manifests) == 0 {
+		return nil, fmt.Errorf("no manifests found in %s", repo.Name())
 	}
-	// TODO: Give back latest image from GCR
-	return nil, nil
-}
 
-func newAuthorizer(tx http.RoundTripper, manager challenge.Manager) transport.RequestModifier {
-	// TODO: Use credentials for GCR configured by user
-	authHandlers := []auth.AuthenticationHandler{
-		auth.NewTokenHandler(tx, nil, "", "pull"),
-		auth.NewBasicHandler(nil),
+	// Determine the latest by sorting by the uploaded time.
+	manifests := make([]google.ManifestInfo, 0, len(res.Manifests))
+	for _, m := range res.Manifests {
+		manifests = append(manifests, m)
+	}
+	sort.Slice(manifests, func(i, j int) bool {
+		return manifests[i].Uploaded.After(manifests[j].Uploaded)
+	})
+	latest := manifests[0]
+	if len(latest.Tags) == 0 {
+		return nil, fmt.Errorf("no tag is associated to the latest image")
 	}
-	return auth.NewAuthorizer(manager, authHandlers...)
+	return &model.ImageRef{
+		ImageName: *image,
+		// TODO: Enable to specify the tag if multiple tags are associated to an image
```

Comment:
I'm looking into using regexp to specify which tag will be used. It would be nice to add the `tagRegex` field to `ImageWatcher` config under the `.pipe/` directory.

```yaml
apiVersion: pipecd.dev/v1beta1
kind: ImageWatcher
spec:
  targets:
    - image: gcr.io/pipecd/foo
      provider: my-gcr
      filePath: foo/deployment.yaml
      field: $.spec.template.spec.containers[0].image
      tagRegex: ^([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$
```

Alternatively, simply providing `useSemver` or something like that might be nice.

Reply:
This option is needed right now for us whose images' have multiple tags: https://gcr.io/pipecd/server

Reply:
It would be wonderful if you could leverage [pubsub](https://cloud.google.com/container-registry/docs/configuring-notifications) instead of polling every 5 minutes for `ImageWatcher`, but I understand that you would need to do something specific for each registry, since [everything implements this slightly differently](https://github.com/opencontainers/distribution-spec/issues/22#issuecomment-417143466).

Reply:
@jonjohnsonjr Thank you for taking a look at our project! You're quite right. We don't have to be worried about the pull rate limit if we could support the event hooks for cloud-provider-specific. However, we decided not to do that so as to reduce the burden on users and avoid complexity.

FYI: Alternatively, we're looking into a completely different solution to deal with the rate limit: https://github.com/pipe-cd/pipe/issues/1341